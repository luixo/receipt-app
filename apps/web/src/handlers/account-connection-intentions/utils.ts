import { TRPCError } from "@trpc/server";
import type { z } from "zod";

import { ACCOUNT_CONNECTIONS_INTENTIONS } from "~db/consts";
import type { AccountsId, UsersId } from "~db/models";
import type { Database } from "~db/types";
import type { emailSchema } from "~web/handlers/validation";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: z.infer<typeof emailSchema>,
	asUserId: UsersId,
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select(["id", "avatarUrl"])
		.where("email", "=", toEmail.lowercase)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `Account with email "${toEmail.original}" does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail.lowercase)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId),
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `Account with email "${toEmail.original}" is already connected to user "${connectedUser.name}".`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where((eb) =>
			eb.and({
				accountId: targetAccount.id,
				targetAccountId: fromAccountId,
			}),
		)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where((eb) =>
					eb.and({
						ownerAccountId: targetAccount.id,
						id: viceVersaIntention.userId,
					}),
				)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where((eb) =>
					eb.and({
						ownerAccountId: fromAccountId,
						id: asUserId,
					}),
				)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: targetAccount.id,
						targetAccountId: fromAccountId,
					}),
				)
				.executeTakeFirst();
		});
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				createdAt: new Date(),
			})
			.executeTakeFirst();
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: fromAccountId,
						targetAccountId: targetAccount.id,
					}),
				)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await database
				.selectFrom("users")
				.select("name")
				.where("id", "=", existingIntention.userId)
				.executeTakeFirstOrThrow();
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${toEmail.original}" as user "${existingUser.name}".`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${user.name}".`,
			});
			// This is probably a c8 bug
			/* c8 ignore next */
		}
		/* c8 ignore next 2 */
		throw e;
	}
};
