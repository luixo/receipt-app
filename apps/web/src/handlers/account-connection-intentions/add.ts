import { TRPCError } from "@trpc/server";
import { entries } from "remeda";
import { z } from "zod";

import type { AccountsId } from "~db/models";
import { queueCallFactory } from "~web/handlers/batch";
import type { AuthorizedContext } from "~web/handlers/context";
import { authProcedure } from "~web/handlers/trpc";
import { emailSchema, userIdSchema } from "~web/handlers/validation";

const addConnectionIntentionSchema = z.strictObject({
	userId: userIdSchema,
	email: emailSchema,
});
type ConnectionIntention = z.infer<typeof addConnectionIntentionSchema>;

const getTargetUsers = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) =>
	ctx.database
		.selectFrom("users")
		.where(
			"users.id",
			"in",
			intentions.map((intention) => intention.userId),
		)
		.leftJoin("accounts", (qb) =>
			qb.onRef("accounts.id", "=", "users.connectedAccountId"),
		)
		.select([
			"users.id",
			"users.name",
			"accounts.email",
			"users.ownerAccountId",
		])
		.execute();

const getTargetAccounts = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) =>
	ctx.database
		.selectFrom("accounts")
		.where(
			"accounts.email",
			"in",
			intentions.map((intention) => intention.email.lowercase),
		)
		.leftJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", ctx.auth.accountId),
		)
		.select([
			"accounts.id",
			"accounts.avatarUrl",
			"accounts.email",
			"users.name as userName",
		])
		.execute();

const getDirectIntentions = async (
	ctx: AuthorizedContext,
	targetAccountsPromise: ReturnType<typeof getTargetAccounts>,
	targetUsersPromise: ReturnType<typeof getTargetUsers>,
) => {
	const targetAccounts = await targetAccountsPromise;
	const targetUsers = await targetUsersPromise;
	if (targetAccounts.length === 0 || targetUsers.length === 0) {
		return [];
	}
	return ctx.database
		.selectFrom("accountConnectionsIntentions")
		.where("accountConnectionsIntentions.accountId", "=", ctx.auth.accountId)
		.where((qb) =>
			qb.or([
				qb(
					"accountConnectionsIntentions.targetAccountId",
					"in",
					targetAccounts.map(({ id }) => id),
				),
				qb(
					"accountConnectionsIntentions.userId",
					"in",
					targetUsers.map(({ id }) => id),
				),
			]),
		)
		.innerJoin("users", (qb) =>
			qb.onRef("users.id", "=", "accountConnectionsIntentions.userId"),
		)
		.select([
			"accountConnectionsIntentions.userId",
			"accountConnectionsIntentions.targetAccountId",
			"users.name",
		])
		.execute();
};

const getViceVersaIntentions = async (
	ctx: AuthorizedContext,
	targetAccountsPromise: ReturnType<typeof getTargetAccounts>,
) => {
	const targetAccounts = await targetAccountsPromise;
	if (targetAccounts.length === 0) {
		return [];
	}
	return ctx.database
		.selectFrom("accountConnectionsIntentions")
		.where(
			"accountConnectionsIntentions.accountId",
			"in",
			targetAccounts.map(({ id }) => id),
		)
		.where(
			"accountConnectionsIntentions.targetAccountId",
			"=",
			ctx.auth.accountId,
		)
		.select([
			"accountConnectionsIntentions.userId",
			"accountConnectionsIntentions.accountId",
		])
		.execute();
};

const getData = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) => {
	const targetAccountsPromise = getTargetAccounts(ctx, intentions);
	const targetUsersPromise = getTargetUsers(ctx, intentions);

	return {
		targetAccounts: await targetAccountsPromise,
		targetUsers: await targetUsersPromise,
		directIntentions: await getDirectIntentions(
			ctx,
			targetAccountsPromise,
			targetUsersPromise,
		),
		viceVersaIntentions: await getViceVersaIntentions(
			ctx,
			targetAccountsPromise,
		),
	};
};

const getIntentionsOrErrors = (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
	{
		targetAccounts,
		targetUsers,
		viceVersaIntentions,
		directIntentions,
	}: Awaited<ReturnType<typeof getData>>,
) =>
	intentions.map((intention) => {
		const targetUser = targetUsers.find((user) => user.id === intention.userId);
		if (!targetUser) {
			return new TRPCError({
				code: "NOT_FOUND",
				message: `User "${intention.userId}" does not exist.`,
			});
		}
		if (targetUser.ownerAccountId !== ctx.auth.accountId) {
			return new TRPCError({
				code: "FORBIDDEN",
				message: `User "${intention.userId}" is not owned by "${ctx.auth.email}".`,
			});
		}
		if (targetUser.email) {
			return new TRPCError({
				code: "CONFLICT",
				message: `User "${intention.userId}" is already connected to account "${targetUser.email}".`,
			});
		}
		const targetAccount = targetAccounts.find(
			(account) => account.email === intention.email.lowercase,
		);
		if (!targetAccount) {
			return new TRPCError({
				code: "NOT_FOUND",
				message: `Account with email "${intention.email.original}" does not exist.`,
			});
		}
		if (targetAccount.userName) {
			return new TRPCError({
				code: "CONFLICT",
				message: `Account with email "${intention.email.original}" is already connected to user "${targetAccount.userName}".`,
			});
		}
		const directIntentionByAccount = directIntentions.find(
			({ targetAccountId }) => targetAccountId === targetAccount.id,
		);
		if (directIntentionByAccount) {
			return new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${intention.email.original}" as user "${directIntentionByAccount.name}".`,
			});
		}
		const directIntentionByUser = directIntentions.find(
			({ userId }) => userId === intention.userId,
		);
		if (directIntentionByUser) {
			return new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${directIntentionByUser.name}".`,
			});
		}
		const viceVersaIntention = viceVersaIntentions.find(
			({ accountId }) => accountId === targetAccount.id,
		);
		if (viceVersaIntention) {
			return {
				type: "vice-versa" as const,
				targetAccount: {
					id: targetAccount.id,
					email: targetAccount.email,
					avatarUrl: targetAccount.avatarUrl,
				},
				asUser: {
					id: intention.userId,
					name: targetUser.name,
				},
				viceVersaUser: {
					id: viceVersaIntention.userId,
				},
			};
		}
		return {
			type: "direct" as const,
			targetAccount: {
				id: targetAccount.id,
				email: targetAccount.email,
				avatarUrl: targetAccount.avatarUrl,
			},
			asUser: {
				id: intention.userId,
				name: targetUser.name,
			},
		};
	});

type Intention = Exclude<
	ReturnType<typeof getIntentionsOrErrors>[number],
	TRPCError
>;

const insertViceVersaIntentions = async (
	ctx: AuthorizedContext,
	intentions: Extract<Intention, { type: "vice-versa" }>[],
) => {
	if (intentions.length === 0) {
		return;
	}
	await ctx.database.transaction().execute((tx) =>
		Promise.all([
			...intentions.map((intention) =>
				tx
					.updateTable("users")
					.set({ connectedAccountId: ctx.auth.accountId })
					.where((eb) =>
						eb.and({
							ownerAccountId: intention.targetAccount.id,
							id: intention.viceVersaUser.id,
						}),
					)
					.executeTakeFirst(),
			),
			...intentions.map((intention) =>
				tx
					.updateTable("users")
					.set({ connectedAccountId: intention.targetAccount.id })
					.where((eb) =>
						eb.and({
							ownerAccountId: ctx.auth.accountId,
							id: intention.asUser.id,
						}),
					)
					.executeTakeFirst(),
			),
			tx
				.deleteFrom("accountConnectionsIntentions")
				.where(
					"accountConnectionsIntentions.targetAccountId",
					"=",
					ctx.auth.accountId,
				)
				.where(
					"accountConnectionsIntentions.accountId",
					"in",
					intentions.map((intention) => intention.targetAccount.id),
				)
				.executeTakeFirst(),
		]),
	);
};

const insertDirectIntentions = async (
	ctx: AuthorizedContext,
	intentions: Extract<Intention, { type: "direct" }>[],
) => {
	if (intentions.length === 0) {
		return;
	}
	await ctx.database
		.insertInto("accountConnectionsIntentions")
		.values(
			intentions.map((intention) => ({
				accountId: ctx.auth.accountId,
				targetAccountId: intention.targetAccount.id,
				userId: intention.asUser.id,
				createdAt: new Date(),
			})),
		)
		.execute();
};

const getDuplicates = <T, K extends string>(
	array: readonly T[],
	getKey: (item: T) => K,
) =>
	entries(
		array
			.map(getKey)
			.reduce<Partial<Record<string, number>>>(
				(acc, element) => ({ ...acc, [element]: (acc[element] ?? 0) + 1 }),
				{},
			),
	).filter(([, value]) => value !== 1);

type IntentionOutput = {
	account: {
		id: AccountsId;
		email: string;
		avatarUrl?: string;
	};
	connected: boolean;
	user: {
		name: string;
	};
};

export const batchFn =
	(ctx: AuthorizedContext) =>
	async (
		inputs: readonly z.infer<typeof addConnectionIntentionSchema>[],
	): Promise<(IntentionOutput | TRPCError)[]> => {
		const duplicatedEmails = getDuplicates(
			inputs,
			(intention) => intention.email.lowercase,
		);
		if (duplicatedEmails.length !== 0) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `Expected to have unique emails, got repeating emails: ${duplicatedEmails
					.map(([email, count]) => `"${email}" (${count} times)`)
					.join(", ")}.`,
			});
		}
		const duplicatedUserIds = getDuplicates(
			inputs,
			(intention) => intention.userId,
		);
		if (duplicatedUserIds.length !== 0) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `Expected to have unique user ids, got repeating: ${duplicatedUserIds
					.map(([userId, count]) => `"${userId}" (${count} times)`)
					.join(", ")}.`,
			});
		}
		const data = await getData(ctx, inputs);
		const intentionsOrErrors = getIntentionsOrErrors(ctx, inputs, data);
		const intentions = intentionsOrErrors.filter(
			(
				intentionOrError,
			): intentionOrError is Exclude<typeof intentionOrError, TRPCError> =>
				!(intentionOrError instanceof TRPCError),
		);
		const directIntentions = intentions.filter(
			(intention): intention is Extract<typeof intention, { type: "direct" }> =>
				intention.type === "direct",
		);
		const viceVersaIntentions = intentions.filter(
			(
				intention,
			): intention is Extract<typeof intention, { type: "vice-versa" }> =>
				intention.type === "vice-versa",
		);
		await Promise.all([
			insertDirectIntentions(ctx, directIntentions),
			insertViceVersaIntentions(ctx, viceVersaIntentions),
		]);
		return intentionsOrErrors.map((intentionOrError) => {
			if (intentionOrError instanceof TRPCError) {
				return intentionOrError;
			}
			const matchedDirectIntention = directIntentions.find(
				({ targetAccount }) =>
					targetAccount.id === intentionOrError.targetAccount.id,
			);
			const matchedViceVersaIntention = viceVersaIntentions.find(
				({ targetAccount }) =>
					targetAccount.id === intentionOrError.targetAccount.id,
			);
			/* c8 ignore start */
			if (!matchedViceVersaIntention && !matchedDirectIntention) {
				return new TRPCError({
					code: "INTERNAL_SERVER_ERROR",
					message: `Expected to have an intention in either direct or vice-versa connection list.`,
				});
			}
			/* c8 ignore stop */
			return {
				account: {
					id: intentionOrError.targetAccount.id,
					email: intentionOrError.targetAccount.email,
					avatarUrl: intentionOrError.targetAccount.avatarUrl || undefined,
				},
				connected: Boolean(matchedViceVersaIntention),
				user: {
					name: intentionOrError.asUser.name,
				},
			};
		});
	};

const queueAddConnectionIntenion = queueCallFactory<
	AuthorizedContext,
	z.infer<typeof addConnectionIntentionSchema>,
	IntentionOutput
>(batchFn);

export const procedure = authProcedure
	.input(addConnectionIntentionSchema)
	.mutation(queueAddConnectionIntenion);
