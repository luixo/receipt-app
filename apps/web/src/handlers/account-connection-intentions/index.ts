import { procedure as accept } from "./accept";
import { procedure as add } from "./add";
import { procedure as getAll } from "./get-all";
import { procedure as reject } from "./reject";
import { procedure as remove } from "./remove";

export const router = {
	getAll,
	add,
	remove,
	accept,
	reject,
};
