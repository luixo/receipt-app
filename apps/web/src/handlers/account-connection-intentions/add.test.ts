import { faker } from "@faker-js/faker";
import { TRPCError } from "@trpc/server";
import { describe, expect } from "vitest";

import { createAuthContext } from "~tests/backend/utils/context";
import {
	insertAccount,
	insertAccountConnectionIntention,
	insertAccountWithSession,
	insertConnectedUsers,
	insertUser,
} from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "~tests/backend/utils/expect";
import { test } from "~tests/backend/utils/test";
import { t } from "~web/handlers/trpc";
import { runInBand } from "~web/handlers/utils.test";

import { procedure } from "./add";

const createCaller = t.createCallerFactory(t.router({ procedure }));

describe("accountConnectionIntentions.add", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			createCaller(context).procedure({
				userId: faker.string.uuid(),
				email: faker.internet.email(),
			}),
		);

		describe("userId", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId: "not a valid uuid",
							email: faker.internet.email(),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "userId": Invalid UUID`,
				);
			});
		});

		describe("email", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId: faker.string.uuid(),
							email: "invalid@@mail.org",
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "email": Invalid email address`,
				);
			});
		});

		test("user does not exist", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));
			const fakeUserId = faker.string.uuid();
			await expectTRPCError(
				() =>
					caller.procedure({
						userId: fakeUserId,
						email: faker.internet.email(),
					}),
				"NOT_FOUND",
				`User "${fakeUserId}" does not exist.`,
			);
		});

		test("user is not owned by an account", async ({ ctx }) => {
			const { sessionId, accountId, account } =
				await insertAccountWithSession(ctx);
			await insertUser(ctx, accountId);

			const { id: foreignAccountId } = await insertAccount(ctx);
			const { id: foreignUserId } = await insertUser(ctx, foreignAccountId);

			const caller = createCaller(await createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						userId: foreignUserId,
						email: faker.internet.email(),
					}),
				"FORBIDDEN",
				`User "${foreignUserId}" is not owned by "${account.email}".`,
			);
		});

		test("target account is not registered", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: userId } = await insertUser(ctx, accountId);

			// Verify that other accounts don't affect error
			await insertAccount(ctx);

			const fakeEmail = faker.internet.email();
			const caller = createCaller(await createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						userId,
						email: fakeEmail,
					}),
				"NOT_FOUND",
				`Account with email "${fakeEmail}" does not exist.`,
			);
		});

		describe("email connection intention exceptions", () => {
			test("target user already has a connected account", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId, email: otherEmail } =
					await insertAccount(ctx);
				const [{ id: userId }] = await insertConnectedUsers(ctx, [
					accountId,
					otherAccountId,
				]);
				// Verify that other users don't affect error
				await insertUser(ctx, accountId);

				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId,
							email: faker.internet.email(),
						}),
					"CONFLICT",
					`User "${userId}" is already connected to account "${otherEmail}".`,
				);
			});

			test("target email is already connected as another user", async ({
				ctx,
			}) => {
				const { accountId, sessionId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId, email: otherEmail } =
					await insertAccount(ctx);
				const [{ name: userName }] = await insertConnectedUsers(ctx, [
					accountId,
					otherAccountId,
				]);

				const { id: userId } = await insertUser(ctx, accountId);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId,
							email: otherEmail,
						}),
					"CONFLICT",
					`Account with email "${otherEmail}" is already connected to user "${userName}".`,
				);
			});

			test("target user already has an intention", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId } = await insertAccount(ctx);
				const { email: targetEmail } = await insertAccount(ctx);
				const { id: userId, name: userName } = await insertUser(ctx, accountId);
				// Self account's intention to connect to foreign account
				await insertAccountConnectionIntention(
					ctx,
					accountId,
					otherAccountId,
					userId,
				);

				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId,
							email: targetEmail,
						}),
					"CONFLICT",
					`You already has intention to connect to user "${userName}".`,
				);
			});

			test("target account already has an intention", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId, email: otherEmail } =
					await insertAccount(ctx);
				const { id: previousUserId, name: userName } = await insertUser(
					ctx,
					accountId,
				);
				// Self account's intention to connect to foreign account
				await insertAccountConnectionIntention(
					ctx,
					accountId,
					otherAccountId,
					previousUserId,
				);

				const { id: userId } = await insertUser(ctx, accountId);

				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							userId,
							email: otherEmail,
						}),
					"CONFLICT",
					`You already has intention to connect to "${otherEmail}" as user "${userName}".`,
				);
			});
		});

		describe("multiple intentions", () => {
			test("duplicate emails", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));

				const { email: otherEmail } = await insertAccount(ctx);

				const { id: userId } = await insertUser(ctx, accountId);
				const { id: anotherUserId } = await insertUser(ctx, accountId);

				await expectTRPCError(
					() =>
						runInBand([
							() => caller.procedure({ userId, email: otherEmail }),
							() =>
								caller.procedure({
									userId: anotherUserId,
									email: otherEmail,
								}),
						]),
					"CONFLICT",
					`Expected to have unique emails, got repeating emails: "${otherEmail}" (2 times).`,
				);
			});

			test("duplicate user ids", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));

				const { email: otherEmail } = await insertAccount(ctx);
				const { email: anotherEmail } = await insertAccount(ctx);

				const { id: userId } = await insertUser(ctx, accountId);

				await expectTRPCError(
					() =>
						runInBand([
							() => caller.procedure({ userId, email: otherEmail }),
							() => caller.procedure({ userId, email: anotherEmail }),
						]),
					"CONFLICT",
					`Expected to have unique user ids, got repeating: "${userId}" (2 times).`,
				);
			});

			test("mixed success and fail", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));

				const {
					email: otherEmail,
					id: otherAccountId,
					avatarUrl: otherAvatarUrl,
				} = await insertAccount(ctx);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const results = await expectDatabaseDiffSnapshot(ctx, () =>
					runInBand([
						() => caller.procedure({ userId, email: otherEmail }),
						() =>
							caller
								.procedure({
									userId: "not a valid uuid",
									email: faker.internet.email(),
								})
								.catch((e) => e),
					]),
				);

				expect(results[0]).toStrictEqual<(typeof results)[0]>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: otherAvatarUrl,
					},
					connected: false,
					user: { name: userName },
				});
				expect(results[1]).toBeInstanceOf(TRPCError);
			});
		});
	});

	describe("functionality", () => {
		describe("account connection intention collapse - has a vice versa intention", () => {
			test("avatar url exists", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const {
					id: otherAccountId,
					email: otherEmail,
					avatarUrl: otherAvatarUrl,
				} = await insertAccount(ctx);

				const { id: otherUserId } = await insertUser(ctx, otherAccountId);
				// Foreign account's intention to connect to self account
				await insertAccountConnectionIntention(
					ctx,
					otherAccountId,
					accountId,
					otherUserId,
				);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const caller = createCaller(await createAuthContext(ctx, sessionId));
				const result = await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({ userId, email: otherEmail }),
				);
				expect(result).toStrictEqual<typeof result>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: otherAvatarUrl,
					},
					connected: true,
					user: { name: userName },
				});
			});

			test("avatar url does not exist", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
					{ avatarUrl: null },
				);

				const { id: otherUserId } = await insertUser(ctx, otherAccountId);
				// Foreign account's intention to connect to self account
				await insertAccountConnectionIntention(
					ctx,
					otherAccountId,
					accountId,
					otherUserId,
				);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const caller = createCaller(await createAuthContext(ctx, sessionId));
				const result = await caller.procedure({ userId, email: otherEmail });
				expect(result).toStrictEqual<typeof result>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: undefined,
					},
					connected: true,
					user: { name: userName },
				});
			});
		});

		describe("account connection intention added", () => {
			test("avatar url exists", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const {
					email: otherEmail,
					id: otherAccountId,
					avatarUrl: otherAvatarUrl,
				} = await insertAccount(ctx);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const caller = createCaller(await createAuthContext(ctx, sessionId));
				const result = await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({ userId, email: otherEmail }),
				);
				expect(result).toStrictEqual<typeof result>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: otherAvatarUrl,
					},
					connected: false,
					user: { name: userName },
				});
			});

			test("avatar url does not exist", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { email: otherEmail, id: otherAccountId } = await insertAccount(
					ctx,
					{ avatarUrl: null },
				);

				const { id: userId, name: userName } = await insertUser(ctx, accountId);

				const caller = createCaller(await createAuthContext(ctx, sessionId));
				const result = await caller.procedure({ userId, email: otherEmail });
				expect(result).toStrictEqual<typeof result>({
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: undefined,
					},
					connected: false,
					user: { name: userName },
				});
			});
		});

		test("multiple intentions", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);

			const { id: otherAccountId, email: otherEmail } = await insertAccount(
				ctx,
				{ avatarUrl: null },
			);

			const { id: otherUserId } = await insertUser(ctx, otherAccountId);
			// Foreign account's intention to connect to self account
			await insertAccountConnectionIntention(
				ctx,
				otherAccountId,
				accountId,
				otherUserId,
			);

			const { id: userId, name: userName } = await insertUser(ctx, accountId);

			const { email: anotherEmail, id: anotherAccountId } = await insertAccount(
				ctx,
				{ avatarUrl: null },
			);

			const { id: anotherUserId, name: anotherUserName } = await insertUser(
				ctx,
				accountId,
			);

			const caller = createCaller(await createAuthContext(ctx, sessionId));
			const results = await runInBand([
				() => caller.procedure({ userId, email: otherEmail }),
				() => caller.procedure({ userId: anotherUserId, email: anotherEmail }),
			]);
			expect(results).toStrictEqual<typeof results>([
				{
					account: {
						id: otherAccountId,
						email: otherEmail,
						avatarUrl: undefined,
					},
					connected: true,
					user: { name: userName },
				},
				{
					account: {
						id: anotherAccountId,
						email: anotherEmail,
						avatarUrl: undefined,
					},
					connected: false,
					user: { name: anotherUserName },
				},
			]);
		});
	});
});
