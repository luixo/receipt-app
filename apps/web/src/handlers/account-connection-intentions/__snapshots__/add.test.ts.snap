// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`accountConnectionIntentions.add > functionality > account connection intention added > avatar url exists 1`] = `
{
  "added": {
    "accountConnectionsIntentions": {
      "accountId:e20766cf-a87a-422e-8f74-6693283d8aee|targetAccountId:4fc3d460-cdcc-43f1-8f2d-46948fa296cc": {
        "accountId": "e20766cf-a87a-422e-8f74-6693283d8aee",
        "createdAt": 2020-01-01T00:00:00.000Z,
        "targetAccountId": "4fc3d460-cdcc-43f1-8f2d-46948fa296cc",
        "updatedAt": 2020-01-01T00:00:00.000Z,
        "userId": "9a11d224-0baf-49c1-89f7-73673387cb63",
      },
    },
  },
}
`;

exports[`accountConnectionIntentions.add > functionality > account connection intention collapse - has a vice versa intention > avatar url exists 1`] = `
{
  "deleted": {
    "accountConnectionsIntentions": {
      "accountId:4fc3d460-cdcc-43f1-8f2d-46948fa296cc|targetAccountId:e20766cf-a87a-422e-8f74-6693283d8aee": undefined,
    },
  },
  "updated": {
    "users": {
      "id:05723e6e-e916-46aa-886e-b744bb05d6ce": {
        "connectedAccountId": "4fc3d460-cdcc-43f1-8f2d-46948fa296cc",
        "updatedAt": 2020-01-01T00:01:00.000Z,
      },
      "id:9a11d224-0baf-49c1-89f7-73673387cb63": {
        "connectedAccountId": "e20766cf-a87a-422e-8f74-6693283d8aee",
        "updatedAt": 2020-01-01T00:01:00.000Z,
      },
    },
  },
}
`;

exports[`accountConnectionIntentions.add > input verification > multiple intentions > mixed success and fail 1`] = `
{
  "added": {
    "accountConnectionsIntentions": {
      "accountId:f69c89cc-b28f-49ab-8c7b-635f1f9b819c|targetAccountId:25b2a1ea-1061-45bb-96a1-84d2ed1dc4d9": {
        "accountId": "f69c89cc-b28f-49ab-8c7b-635f1f9b819c",
        "createdAt": 2020-01-01T00:00:00.000Z,
        "targetAccountId": "25b2a1ea-1061-45bb-96a1-84d2ed1dc4d9",
        "updatedAt": 2020-01-01T00:00:00.000Z,
        "userId": "20566048-adf0-457d-936e-ecfb1d719980",
      },
    },
  },
}
`;
